Add MV_CONTAINS function

mjmbischoff opened the pull request:
Adding the MV_CONTAINS_ALL function.

The function checks if the multivalue expression that is supplied as the first argument resolves to all values yielded by the multivalue expression that is the second argument. Effectively establishing if the first argument is a super set of the second argument which would then be the subset. This allows for a contains element function that also has the flexibility to work with multivalue arguments. Since we compare each entry duplicates collapse as we simply match against the same entry on the other side.

This avoids the need to abuse string functions to check if something is one of the multi values.:

```
FROM main_data_index
| LOOKUP JOIN my_lookup_index ON destination.ip
| WHERE CONTAINS(CONCAT("#",MV_CONCAT(destination.denied_ports::STRING, "#"),"#"), CONCAT("#",destination.port::STRING,"#"))
| KEEP message, destination.*
```

```
FROM main_data_index
| LOOKUP JOIN my_lookup_index ON destination.ip
| WHERE MV_CONTAINS(destination.denied_ports, destination.port)
| KEEP message, destination.*
```

the function allows for multivalue so the following should also work:
```
MV_CONTAINS(roles, ["read", "write"])
```

---
TODO
- [x] Regenerate docs
- [ ] Rename MV_CONTAINS_ALL to MV_CONTAINS (pending decision - jumped the gun but should check for consensus)
- [x] Introduce as PREVIEW or straight to GA?

mjmbischoff committed:
Add `mv_contains_all` function

Commit:
[CI] Auto commit changes from spotless

mjmbischoff committed:
Enhance `mv_contains_all` function with additional examples and documentation improvements.

Fix issue with byteref being empty, which caused fold to fail.

Commit:
[CI] Auto commit changes from spotless

mjmbischoff committed:
Add documentation for `mv_contains_all` function

mjmbischoff committed:
Fixing documentation for `mv_contains_all` function

mjmbischoff committed:
Regenerated `mv_contains_all` documentation

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

mjmbischoff committed:
Update docs/changelog/133099.yaml

elasticsearchmachine commented:
Hi @mjmbischoff, I've created a changelog YAML for you.

mjmbischoff committed:
Update docs/changelog/133099.yaml

mjmbischoff committed:
Merge branch 'main' into esql-MV_CONTAINS_ALL

mjmbischoff committed:
Update docs/changelog/133099.yaml

mjmbischoff committed:
Readding skipped markers as else the test gets run in the backwards compatibility test environment. - not sure how to test it as, I feel like the version should be main on main / dev. Doing the dance for now.

mjmbischoff committed:
Merge branch 'main' into esql-MV_CONTAINS_ALL

mjmbischoff committed:
Merge branch 'main' into esql-MV_CONTAINS_ALL

leemthompo commented on docs/changelog/133099.yaml:
```diff
@@ -0,0 +1,5 @@
+pr: 133099
+summary: Add MV_CONTAINS_ALL ES:QL function
```
```suggestion
summary: Add MV_CONTAINS_ALL function
```

leemthompo commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvContainsAll.java:
```diff
@@ -0,0 +1,283 @@
+    );
+    private DataType dataType;
+
+    @FunctionInfo(
```
You'll need to add applies_to info here so we can tag what version this was introduced in the docs and whether it's tech preview or GA in 9.2. See the [docs README](https://github.com/elastic/elasticsearch/blob/main/docs/reference/query-languages/esql/README.md#version-differentiation-in-docs-v3) for more info.

mjmbischoff replied:
resolved in https://github.com/elastic/elasticsearch/pull/133099/commits/23544d4bba520d40de057836c8ff8dbe25fd7c93

leemthompo reviewed (commented):
Need to make sure the docs are tagged for version and lifecycle :)

leemthompo commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvContainsAll.java:
```diff
@@ -0,0 +1,283 @@
+    @FunctionInfo(
+        returnType = "boolean",
+        description = "Checks if the values yielded by multivalue value expression are all present in the values yielded by the other "
+            + "multivalue expression. The result is a boolean representing the outcome or null if either of the expressions where null.",
```
I might actually rewrite this to something like this:


_"Checks if all values yielded by the second multivalue expression are present in the values yielded by the first multivalue expression. Returns a boolean, or null if either expression is null."_

mjmbischoff replied:
resolved in https://github.com/elastic/elasticsearch/pull/133099/commits/23544d4bba520d40de057836c8ff8dbe25fd7c93

leemthompo replied:
@mjmbischoff you'll need to regenerate the docs too :)

mjmbischoff committed:
Update `mv_contains_all` docs to reflect the correct version and add lifecycle metadata
Documentation rewording.

Co-authored-by: Liam Thompson <leemthompo@gmail.com>

mjmbischoff commented:
> Need to make sure the docs are tagged for version and lifecycle :)

Not sure why I though 9.2 was already out the door but fixed 9.3->9.2 as well. thx!

Commit:
[CI] Auto commit changes from spotless

mjmbischoff committed:
Merge branch 'main' into esql-MV_CONTAINS_ALL

leemthompo commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvContainsAll.java:
```diff
@@ -0,0 +1,286 @@
+            @Example(file = "string", tag = "mv_contains_all"),
+            @Example(file = "string", tag = "mv_contains_all_bothsides"),
+            @Example(file = "string", tag = "mv_contains_all_where"), },
+        appliesTo = { @FunctionAppliesTo(lifeCycle = FunctionAppliesToLifecycle.PREVIEW, version = "9.2.0") }
```
FYI the docs need to be regenerated here, they are currently stale versus later commits

this is correct if the function is going to be released as tech preview

should be `FunctionAppliesToLifecycle.GA` if it's going straight to GA :)

leemthompo replied:
haha you still need to regenerate the docs one way or another so yeah start preview and we can change after ;)

until 9.2 actually releases the rendered tag will just say "Planned" anyways so we have until release day to decide

craigtaverner commented:
> A bit on the fence on the argument order, if we want to introduce this as a operator: given `IN MV`, `WHERE port IN MV allowed_ports` or `port =c allow_ports`. Which would be reverse order. I went with superset, subset as java `superset.containsAll(subset)`.

The verbs `CONTAINS` and `IN` (or `WITHIN`) are converse of each other. I think since the function here has the word `CONTAINS` the parameter order should be superset, subset. If the function instead had the word `IN` or `WITHIN`, then the parameter order should be subset, superset. This is similar to the existing spatial functions `[ST_CONTAINS](https://www.elastic.co/docs/reference/query-languages/esql/functions-operators/spatial-functions#esql-st_contains)` and `[ST_WITHIN](https://www.elastic.co/docs/reference/query-languages/esql/functions-operators/spatial-functions#esql-st_within)`.

mjmbischoff committed:
Processing review, changing null handling

Commit:
[CI] Auto commit changes from spotless

mjmbischoff committed:
Slightly more efficient null handling and fix null test logic

Commit:
[CI] Auto commit changes from spotless

mjmbischoff committed:
Merge branch 'main' into esql-MV_CONTAINS_ALL

mjmbischoff committed:
- Renamed MV_CONTAINS_ALL to MV_CONTAINS
- Fixing tests by removing logic to return null if all parameters are null. The standard generator had to be circumvented, should follow up with separate PR to make it more intelligent to avoid it.
- Overwritten part of the test methods to avoid the null expectation.

Commit:
[CI] Auto commit changes from spotless

mjmbischoff committed:
Update documentation to reflect `MV_CONTAINS_ALL` renaming to `MV_CONTAINS`

mjmbischoff commented:
Had to do quite a bit(too much?) to follow the contract of always returning a boolean vs null if any argument is null, which the test classes seem to expect.

ivancea reviewed (approved):
It looks good! It's a shame you had to create a custom evaluator; I guess we'll have to invest some extra time there, as MV functions can usually handle nulls

mjmbischoff committed:
Fix `MvContains` javadoc

mjmbischoff committed:
Add preview marker to documentation link

Co-authored-by: Iván Cea Fontenla <ivancea96@outlook.com>

mjmbischoff commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvContainsTests.java:
```diff
@@ -0,0 +1,368 @@
+    // We always return a boolean.
+    @Override
+    protected Matcher<Object> allNullsMatcher() {
+        return anyOf(equalTo(false), equalTo(true));
```
Naming is poor here as this matcher/method is used in cases where any argument is null. See AbstractScalarFunctionTestCase#testEvaluateBlock(). Same as with the other issue should fix this in a separate PR.

mjmbischoff committed:
I think checkstyle needs to be updated :D

mjmbischoff committed:
Merge branch 'main' into esql-MV_CONTAINS_ALL

mjmbischoff committed:
Merge branch 'main' into esql-MV_CONTAINS_ALL

mjmbischoff committed:
Replacing table and markdown comment block

mjmbischoff committed:
Merge branch 'main' into esql-MV_CONTAINS_ALL

mjmbischoff committed:
Merge branch 'main' into esql-MV_CONTAINS_ALL
